### DISCUSSION TITLE
Guidance on manually implementing `Any`, getters etc

### DISCUSSION THREAD

**Discussion by VorpalBlade** [rune-rs/rune#727] [Q&A] [answered]
I have a type that looks like:

```rust
pub struct PciDeviceVendorLookup<S> {
    pub vendor: Option<S>,
    pub device: Option<S>,
    pub subvendor: Option<S>,
    pub subdevice: Option<S>,
}
```

This gets parameterised either on `&'lifetime str` when used from Rust code (where borrowing works) or `String` when exposed to a rune script. I also have a `to_owned` to convert from the borrow to the owned version. I can't not derive `Any` for this:

```
147 | #[cfg_attr(feature = "rune", derive(rune::Any))]
    |                                     ^^^^^^^^^
    |                                     |
    |                                     the parameter type `S` must be valid for the static lifetime...
    |                                     ...so that the type `S` will meet its required lifetime bounds
```

As such I want to implement `Any` and `rune(get)` on the fields only when `S = String`. This should be perfectly possible to do by hand, something like:

```rust
impl rune::Any for PciVendorLookup<String> {
    fn type_hash() -> rune::Hash {
        todo!()
    }
}

impl rune::compile::Named for PciVendorLookup<String> {
    const BASE_NAME: rune::runtime::RawStr = todo!();
}
```

Unfortunately I can't find any guidance on what to put for `todo!()` in the documentation (maybe I missed it?).

Nor can I find anything much on how to implement associated / inherent functions or getters in this case. My proc-macro-fu is very lacking, I tried to figure out what I'm supposed to do that way, but no luck.

**Comment by VorpalBlade** [comment]
I had a read of what the proc-macro generates and derive Any seems to also derive a bunch of other things. `Named` looks simple enough to implement, but `Any` and `TypeOf` seems to use `doc(hidden)` functions. And I couldn't figure out how the `InstallWith` was supposed to work. I did not look into how `UnsafeToRef`/`UnsafeToMut`/`UnsafeToValue`/... works.

Yes `derive(Any)` seems to be doing a lot of things, and it is unfortunate that it doesn't seem feasible to do it by hand when needed.

 So I don't see a way forward on my own here, apart from creating two non-generic structs (but that would be admitting defeat). Any guidance would be appreciated @udoprog.

**Reply by udoprog** [reply, marked as answer]
This would primarily be about missing bounds over `S`. We actually have a test case for a generic type using `Any`:

https://github.com/rune-rs/rune/blob/38de932d0d71f2b3d89651410cc8757b2bb2aa36/crates/rune/src/tests/external_generic.rs#L18

It would be better if we added the bounds per trait as-needed instead of requiring that it's present over the whole type, since now it would have to be present for every impl block related to the type.

**Reply by VorpalBlade** [reply]
That would be a good way to do it, being able to support rune usage only when the generic parameters are constrained (i.e. have the bounds on the impls, not on the type itself, which is good Rust practise anyway, as can be seen on many types in the standard library).

Especially since in my case it will either be String or &'a str, the latter definitely doesn't meet the bounds and as such the bounds cannot be on the struct itself.
